nomos-services/chain-service/src/bootstrap/ibd.rs

```diff
@@ -1,50 +1,785 @@
-            InitialCryptarchia::Bootstrapping(cryptarchia) => {
-                let cryptarchia = Self::run_with(cryptarchia, network_adapter).await;
-                IbdCompletedCryptarchia::Bootstrapping(cryptarchia)
-            }
-            InitialCryptarchia::Online(cryptarchia) => {
-                let cryptarchia = Self::run_with(cryptarchia, network_adapter).await;
-                IbdCompletedCryptarchia::Online(cryptarchia)
+            InitialCryptarchia::Bootstrapping(cryptarchia) => self
+                .run_with(cryptarchia, network_adapter, block_processor)
+                .await
+                .map(IbdCompletedCryptarchia::Bootstrapping),
+            InitialCryptarchia::Online(cryptarchia) => self
+                .run_with(cryptarchia, network_adapter, block_processor)
+                .await
+                .map(IbdCompletedCryptarchia::Online),
+        }
+    }
+
+    /// Runs the IBD with the provided [`Cryptarchia`].
+    /// This consumes [`Cryptarchia`] and returns the new one after completion.
+    async fn run_with<CryptarchiaState, BlockProcessor>(
+        self,
+        cryptarchia: Cryptarchia<CryptarchiaState>,
+        network_adapter: NetAdapter,
+        block_processor: &mut BlockProcessor,
+    ) -> Result<Cryptarchia<CryptarchiaState>, IbdError>
+    where
+        CryptarchiaState: cryptarchia_engine::CryptarchiaState + Send,
+        BlockProcessor: processor::BlockProcessor<Block = NetAdapter::Block> + Send + Sync,
+    {
+        // TODO: Download from multiple peers.
+        //       https://github.com/logos-co/nomos/issues/1455
+        match self.config.peers.iter().next() {
+            Some(peer) => self
+                .download_from_peer(cryptarchia, network_adapter, *peer, block_processor)
+                .await
+                .map_err(|(_, e)| e),
+            None => Ok(cryptarchia),
+        }
+    }
+
+    /// Downloads blocks from a peer until its tip is reached.
+    /// This returns the updated [`Cryptarchia`] after downloading blocks.
+    /// If any error occurs during the download,
+    /// it returns [`Cryptarchia`] that contains blocks processed before the
+    /// error, so that the caller can retry the download.
+    async fn download_from_peer<CryptarchiaState, BlockProcessor>(
+        &self,
+        mut cryptarchia: Cryptarchia<CryptarchiaState>,
+        network_adapter: NetAdapter,
+        peer: NetAdapter::PeerId,
+        block_processor: &mut BlockProcessor,
+    ) -> Result<Cryptarchia<CryptarchiaState>, (Cryptarchia<CryptarchiaState>, IbdError)>
+    where
+        CryptarchiaState: cryptarchia_engine::CryptarchiaState + Send,
+        BlockProcessor: processor::BlockProcessor<Block = NetAdapter::Block> + Send + Sync,
+    {
+        // Track this to avoid downloading the same block multiple times.
+        let mut latest_downloaded_block: Option<HeaderId> = None;
+
+        // Repeat until the peer's tip is reached.
+        loop {
```

Comment:
should we have some stop gap solution like a max number of attempts?

Reply:
they could force you to wait forever in the bootstrap phase by not giving you any blocks